### PR TITLE
(fix): update the forwarder to use a thread

### PR DESCRIPTION
**Pull request by thomaszurkan-optimizely**
## Summary
- Update the event forwarder to use a thread instead of using the main thread.

## Test plan
- update tests

### CODE REVIEW

File: lib/optimizely/event/forwarding_event_processor.rb [hunk 1]
```diff
@@ -30,11 +30,14 @@ def process(user_event)
       log_event = Optimizely::EventFactory.create_log_event(user_event, @logger)
 
       begin
-        @event_dispatcher.dispatch_event(log_event)
-        @notification_center&.send_notifications(
-          NotificationCenter::NOTIFICATION_TYPES[:LOG_EVENT],
-          log_event
-        )
+        @notification_center&.send_notifications(NotificationCenter::NOTIFICATION_TYPES[:LOG_EVENT], log_event)
```
Comment (oakbani) [inline lib/optimizely/event/forwarding_event_processor.rb]: Shouldn't we be only firing notification when the event has successfully dispatched?

File: lib/optimizely/event/forwarding_event_processor.rb [hunk 2]
```diff
@@ -30,11 +30,14 @@ def process(user_event)
...
+        Thread.new do
+          begin
+            @event_dispatcher.dispatch_event(log_event)
+          rescue StandardError => e
```
Comment (oakbani) [inline lib/optimizely/event/forwarding_event_processor.rb]: Can we run into second rescue when we have rescued here?

File: spec/project_spec.rb [hunk 3]
```diff
@@ -492,10 +497,12 @@ class InvalidErrorHandler; end
       project_instance.decision_service.set_forced_variation(project_config, 'test_experiment_with_audience', 'test_user', 'variation_with_audience')
       variation_to_return = project_instance.decision_service.get_forced_variation(project_config, 'test_experiment', 'test_user')
       allow(project_instance.decision_service.bucketer).to receive(:bucket).and_return(variation_to_return)
+      sleep 0.1
```
Comment (oakbani) [inline spec/project_spec.rb]: I am not sure if we need to put a sleep before **allow**.

File: spec/project_spec.rb [hunk 4]
```diff
@@ -595,13 +602,16 @@ def callback(_args); end
         method(:callback)
       )
       variation_to_return = project_config.get_variation_from_id('test_experiment', '111128')
+      sleep 0.1
```
Comment (oakbani) [inline spec/project_spec.rb]: I am not sure if we need to put a sleep before allow.

File: spec/project_spec.rb [hunk 5]
```diff
@@ -741,18 +757,20 @@ def callback(_args); end
 
         expect(http_project_config_manager.config).not_to eq(nil)
         expect(project_instance.activate('checkout_flow_experiment', 'test_user')).not_to eq(nil)
+        sleep 0.5
```
Comment (oakbani) [inline spec/project_spec.rb]: why sleeping for 0.5 here? I fear flakiness in  Unit Tests on Travis.

File: spec/project_spec.rb [hunk 6]
```diff
@@ -2814,7 +2852,7 @@ def callback(_args); end
       project_instance = Optimizely::Project.new(nil, nil, nil, nil, true, nil, nil, config_manager, nil, event_processor)
 
       expect(config_manager.stopped).to be false
-      expect(event_processor.started).to be true
+      expect(event_processor.started).to be false
```
Comment (oakbani) [inline spec/project_spec.rb]: doesn't belong to this PR?

File: lib/optimizely/event/forwarding_event_processor.rb (hunk 1 above)
Comment (oakbani) [reply lib/optimizely/event/forwarding_event_processor.rb]: Also, why not also fire notification in the thread as well. this will allow us to quickly free up our main thread if the client has registered some time-taking callback in the LogEvent listener. And, if we do this, we can simply wait for the callback to get called in our unit tests rather than putting it to sleep before assertions.

Comment (thomaszurkan-optimizely) [reply lib/optimizely/event/forwarding_event_processor.rb]: Yeah, I was debating back and forth on that.  Originally, it was in the thread.  However, all the other SDKs fire the event before dispatch.  So, I thought this was ok.  I can put it back in the thread.  it might be better for the notification listener to not be on the main thread.

File: lib/optimizely/event/forwarding_event_processor.rb (hunk 2 above)
Comment (thomaszurkan-optimizely) [reply lib/optimizely/event/forwarding_event_processor.rb]: The first rescue is for the event dispatch failure.  really just for logging.  the second is for thread failure.

File: spec/project_spec.rb (hunk 3 above)
Comment (thomaszurkan-optimizely) [reply spec/project_spec.rb]: there is a get_forced _varitation before that.  the sleep is to give the thread a chance to run.

File: spec/project_spec.rb (hunk 4 above)
Comment (thomaszurkan-optimizely) [reply spec/project_spec.rb]: the thread will not have a chance to run without the sleep.

**Comment by aliabbasrizvi** [comment]
Closing due to inactivity. Please re-open if needed.
